### PR TITLE
Add filter method to Array class

### PR DESCRIPTION
**Pull request by AlexDarigan**
Designed to be consistent with similar methods such as
sort_custom instead of using FuncRef.

The error isn't always accurate, it is sometimes due
to the wrong amount of parameters defined in the filterer method
(which should be 2, one for the array element, another for other args)

### CODE REVIEW

**Comment by fire** [comment]
Are you familiar with this pr? https://github.com/godotengine/godot/pull/20545

**Comment by CodeDarigan** [comment]
I am. I decided to make a new PR because I'm not entirely familler with the workflow of PRs and I also felt like this would be best dealt with on a PR per method basis.

**Comment by CodeDarigan** [comment]
Looking at the array class, I see sort, sort_custom and invert don't return a list when being called in GDScript. The typical implementation of filter returns a new value (as far as I know). Would it be more prudent to be consistent with other GDScript methods or the typical implementation of functional methods? (in-place vs return a copy)

**Comment by sparkart** [comment]
The other PR seems to have alot of work done and the only problem with it at the moment is merge conflicts. Its probably better to just resolve its conflicts.

File: core/array.cpp [hunk 1]
```diff
@@ -335,6 +335,26 @@ Array &Array::invert() {
...
+	for (int i = 0; i < size(); i++) {
+
+		Variant res = p_obj->call(p_function, get(i), p_args);
+
+		if (res.get_type() != Variant::Type::BOOL) {
+			ERR_EXPLAIN("Comparison function must return a value of type bool");
+			ERR_FAIL_V(ret);
+		}
+
+		if (((bool)res)) {
```
Comment (akien-mga) [inline core/array.cpp]: There's one pair of parentheses too many.

**Comment by akien-mga** [comment]
The commits should be squashed together. Otherwise it seems OK, though it would be nice to find how to solve this:

> The error isn't always accurate, it is sometimes due
to the wrong amount of parameters defined in the filterer method
(which should be 2, one for the array element, another for other args)

(cc @bojidar-bg)

File: core/array.cpp (hunk 1 above)
Comment (bojidar-bg) [reply core/array.cpp]: We use `.operator bool()` in other places to convert from `Variant` to `boolean`. There is also `.booleanize()` AFAIK.

**Comment by bojidar-bg** [comment]
> The error isn't always accurate, it is sometimes due to the wrong amount of parameters defined in the filterer method (which should be 2, one for the array element, another for other args)

Just use `Object::call(const StringName&, const Variant**, int, Variant::CallError&)`, and display the call error if it is not `CALL_OK`.

I would suggest either turning `filter` into a vararg method, changing `args` to an actual array (similar to binds), or removing `args` altogether.

File: core/array.cpp (hunk 1 above)
Comment (CodeDarigan) [reply core/array.cpp]: Resolved in #eedfaea

**Comment by CodeDarigan** [comment]
@akien-mga 7bc675e should have squashed all current commits.

@bojidar-bg I've been trying a few things out but never succeeding so I think I'm out of my depth here and I don't really have the time to spend learning more about it. 

For the record, the issue itself wasn't with filter but the user-defined method and I don't *think* there is a way to do that that works in all cases without invoking an error from the c++ side or a warning from the gdscript side (if the arg is redundant or unused) and I think removing the arg array sort of makes filter pointless if it is going to be a strict condition check over a parameterized one.

As I said though, I might just be out of my depth.

**Comment by vnen** [comment]
Honestly, I'd wait until we have first-class functions (which should be soon) to add this and other higher-order functions. It'll be better for regular usage and I'd rather wait than break the API later.

**Comment by CodeDarigan** [comment]
Yeah. That would probably be for the best.

**Comment by KoBeWi** [comment]
I was waiting for it to be merged to implement other methods :<

**Pull request closed without merging**